### PR TITLE
[gms-955][feature] eip 4494 permits for hybrid and mint by id preset contracts

### PR DESCRIPTION
**Pull request by jasonzwli**
https://eips.ethereum.org/EIPS/eip-1271
https://eips.ethereum.org/EIPS/eip-4494
https://eips.ethereum.org/EIPS/eip-712

### CODE REVIEW

File: contracts/token/erc721/abstract/ERC721Permit.sol [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        address spender,
+        uint256 tokenId,
+        uint256 deadline,
+        bytes memory sig
+    ) external override {
+        _permit(spender, tokenId, deadline, sig);
+    }
+
+    /**
+     * @notice [ERC-4494] Function to approve by way of owner signature
```
Comment (ipekt) [inline contracts/token/erc721/abstract/ERC721Permit.sol:60]:
```suggestion
     * @notice [ERC-4494] Approve given token for given spender by owner signature
```

File: contracts/token/erc721/abstract/ERC721Permit.sol [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+     * @param spender the address to approve
+     * @param tokenId the index of the NFT to approve the spender on
```
Comment (ipekt) [inline contracts/token/erc721/abstract/ERC721Permit.sol:62]:
```suggestion
     * @param spender the spender address to approve
     * @param tokenId the ID of the NFT to approve the spender on
```

**Review by ipekt — DISMISSED** [review]

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                digest,
+                bytes32(BytesLib.slice(sig, 0, 32)),
+                bytes32(BytesLib.slice(sig, 32, 64))
+            );
+        } else if (sig.length == 65) {
+            // typical EDCSA Sig
+            recoveredSigner = ECDSA.recover(digest, sig);
+        } else {
+            revert InvalidSignature();
+        }
```
Comment (lfportal) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:84]: **Question:** Would this `revert` for an `ERC-1271` based signature? My understanding is `ERC-1271` makes no assumptions about signatures and allows arbitrary length.

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+        if (!isNotZerothAddr) {
+            revert SignerCannotBeZerothAddress();
+        }
+
+        bool isApprovedOperator = _isApprovedOrOwner(recoveredSigner, tokenId);
+
+        bool isValidEOASig = isApprovedOperator && isNotZerothAddr;
+
+        if (!isValidEOASig) {
```
Comment (valzam) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:96]: This is also hard to parse, what about

```
        if (
            _isApprovedOrOwner(recoveredSigner, tokenId) ||
            (_isValidERC1271Signature(getApproved(tokenId), digest, sig) ||
                _isValidERC1271Signature(ownerOf(tokenId), digest, sig))
        ) {
            _approve(spender, tokenId);
        }

        revert InvalidSignature();
```

I find this easier to understand because the boolean conditions aren't all `not`-ed. It also by default reverts, which I think would be preferable from a security perspective.

Sidenote, why are we checking isApprovedOrOwner explicitly? Just as an optimisation to not do signature verification if the owner/operator calls this permit function? In that case, our default use case is exchanges, which will not be owner/approved, would it be more gas efficient to just go straight to signature verification?

File: contracts/token/erc721/abstract/ERC721Permit.sol [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    /**
+     * @notice Checks if a given signature is valid according to EIP-1271.
+     * @param spender The address which purports to have signed the message.
+     * @param digest The EIP-712 compliant digest that was signed.
+     * @param sig The actual signature bytes.
+     * @return True if the signature is valid according to EIP-1271, otherwise false.
+     */
+    function _isValidERC1271Signature(address spender, bytes32 digest, bytes memory sig) private view returns(bool) {
```
Comment (valzam) [inline contracts/token/erc721/abstract/ERC721Permit.sol:165]: Did you base this on an existing implementation or is there a doc that describes the logic here?

**Review by valzam — COMMENTED** [review]
Let's have a chat about this, I left some questions that I would like to discuss.

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+
+        if (recoveredSigner == address(0)) {
+            revert SignerCannotBeZerothAddress();
+        }
+
+        if (
+            _isApprovedOrOwner(recoveredSigner, tokenId) ||
+                _isValidERC1271Signature(getApproved(tokenId), digest, sig) ||
+                _isValidERC1271Signature(ownerOf(tokenId), digest, sig)
+        ) {
```
Comment (lfportal) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:94]: **Question:** It isn't clear to me why an approved operator should be allowed to create permits on behalf of an owner. I also haven't found this to be specified in the ERC-4494 specification. Is this necessary?

If an approved operator is allowed to create permits, we'd also need to get for each operator approved by `setApprovalForAll` when performing ERC-1271 sig checks. I'm not sure if there's a good way to do this.

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+            revert InvalidSignature();
+        }
+
+        if (recoveredSigner == address(0)) {
+            revert SignerCannotBeZerothAddress();
+        }
+
+        if (
+            _isApprovedOrOwner(recoveredSigner, tokenId) ||
+                _isValidERC1271Signature(getApproved(tokenId), digest, sig) ||
```
Comment (jasonzwli) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:91]: remove

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol (hunk 3 above)
Comment (jasonzwli) [reply contracts/token/erc721/abstract/ERC721HybridPermit.sol:89]: Yep, moved the eip1271 validation to the top of the method. We still need to pass the digest and sig to the eip1271 wallet as thats the specified interface.

File: contracts/token/erc721/abstract/ERC721Permit.sol [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+    {}
+    
+    /**
+     * @notice Function to approve by way of owner signature
+     * @param spender the address to approve
+     * @param tokenId the index of the NFT to approve the spender on
+     * @param deadline a timestamp expiry for the permit
+     * @param sig a traditional or EIP-2098 signature
+     */
+    function permit(
```
[the remark this answers is not in the document]
Comment (jasonzwli) [reply contracts/token/erc721/abstract/ERC721Permit.sol:51]: Our options are quite limited. As the two contracts have complete independent inheritance trees. 
1. A Library which means an extra contract to manage
2. Not extending erc721 in the permit and calling erc721 related methods via address. This results in the msg.sender being changed to the collection itself, which will trigger allowed operator check problems

File: contracts/token/erc721/abstract/ERC721Permit.sol (hunk 5 above)
Comment (jasonzwli) [reply contracts/token/erc721/abstract/ERC721Permit.sol:163]: The eip itself has an example implementation that calls the method directly on the wallet, however that does not let us revert with our own errors, the staticcall here is to give us more control over what behaviour we want if the signature verification fails within the eip1271 contract

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol (hunk 6 above)
Comment (jasonzwli) [reply contracts/token/erc721/abstract/ERC721HybridPermit.sol]: Removed as an result of offline discussion.

**Review by lfportal — APPROVED** [review]
LGTM!

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+                    tokenId,
+                    _nonces[tokenId],
+                    deadline
+                )
+            )
+        );
+    }
+
+    function _isValidEOASignature(address recoveredSigner, uint256 tokenId) private view returns(bool) {
+        if (recoveredSigner == address(0)) {
```
Comment (valzam) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:143]: can we simplify this to 
```
return recoveredSigner == address(0) && _isApprovedOrOwner(recoveredSigner, tokenId)
```

File: contracts/token/erc721/abstract/ERC721HybridPermit.sol [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        uint256 deadline,
+        bytes memory sig
+    ) internal virtual {
+        if (deadline < block.timestamp) {
+            revert PermitExpired();
+        }
+
+        bytes32 digest = _buildPermitDigest(spender, tokenId, deadline);
+        
+        if (_isValidERC1271Signature(ownerOf(tokenId), digest, sig)) {
```
Comment (valzam) [inline contracts/token/erc721/abstract/ERC721HybridPermit.sol:70]: Let's please add some comments here to signal that the first half deals with SC wallets and the second half with EOA signers.

**Review by Jon-Alonso — APPROVED** [review]

**Review by hysmio — APPROVED** [review]

**Review by valzam — APPROVED** [review]
